ccp/src/hashrate/hashrate_collector.rs

```diff
@@ -0,0 +1,240 @@
+pub(crate) struct CUHashrateRaw {
+    cache_creation: ParameterStatus<Duration>,
+    dataset_initialization: Vec<Duration>,
+    pow_duration: ParameterStatus<Duration>,
+    hashes_checked: u64,
+    proofs_found: u64,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) enum ParameterStatus<T> {
+    Measured(T),
+    NotMeasured,
+}
+
+pub(crate) type Hashrate = HashMap<LogicalCoreId, CUHashrate>;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct CUHashrate {
+    // hashes from epoch start, counts operations with cache and dataset
+    pub(crate) effective_hashrate: f64,
+    // pure hashrate, which doesn't count operations with cache and dataset
+    pub(crate) hashrate: ParameterStatus<f64>,
+    pub(crate) cache_creation: ParameterStatus<Duration>,
+    pub(crate) dataset_initialization: ParameterStatus<Duration>,
+}
+
+impl HashrateCollector {
+    pub(crate) fn new() -> Self {
+        Self {
+            status: CollectorStatus::JustCreated,
+            entries: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn count_entry(&mut self, hashrate_record: HashrateCURecord) -> Option<Hashrate> {
+        use std::collections::hash_map::Entry;
+
+        let result = match self.status {
+            CollectorStatus::JustCreated => {
+                self.new_epoch(hashrate_record.epoch);
+                None
+            }
+            CollectorStatus::Started { epoch, .. } if epoch == hashrate_record.epoch => None,
+            CollectorStatus::Started { .. } => {
+                let statistics = self.collect();
+                self.new_epoch(hashrate_record.epoch);
+
+                Some(statistics)
+            }
+        };
+
+        match self.entries.entry(hashrate_record.core_id) {
+            Entry::Vacant(entry) => {
+                entry.insert(CUHashrateRaw::from_single_record(hashrate_record));
+            }
+            Entry::Occupied(entry) => entry.into_mut().count_record(hashrate_record),
+        };
+
+        result
+    }
+
+    pub(crate) fn collect(&self) -> HashMap<LogicalCoreId, CUHashrate> {
+        let epoch_duration = match self.status {
+            CollectorStatus::Started { started_time, .. } => started_time.elapsed(),
+            CollectorStatus::JustCreated => return HashMap::new(),
+        };
+
+        let epoch_duration = epoch_duration.as_secs_f64();
+
+        self.entries
+            .iter()
+            .map(|(&core_id, info)| {
+                let hashrate = match info.pow_duration {
+                    ParameterStatus::Measured(duration) => {
+                        let hashrate = info.hashes_checked as f64 / duration.as_secs_f64();
+                        ParameterStatus::Measured(hashrate)
+                    }
+                    ParameterStatus::NotMeasured => ParameterStatus::NotMeasured,
+                };
+
+                let dataset_initialization = match info.dataset_initialization.iter().max() {
```

Comment:
`From<Option<T>> for ParameterStatus<T>`?

Reply:
It's made by accident, this field should also has `ParameterStatus` type.